### PR TITLE
Add validation cache

### PR DESCRIPTION
**Pull request by Trejjam**
On multiple validate call, failed

### CODE REVIEW

**Comment by uestla** [comment]
Can you please add more details about your use case?

Also can you please update the changes to follow coding standards?

Thanks!

**Comment by Trejjam** [comment]
Yes, i can. Sometimes i need re-validate form. If you create *nd request for validate captcha key, it will fail, because key was already validated.

I will look on coding standard, thank you for hints.

**Comment by uestla** [comment]
And this re-validation is done by AJAX request or normal POST request? Does this have something to do with live validation?

**Comment by Trejjam** [comment]
In one request by twice calling `$form->validate()` (First is done, second fail)

File: src/ReCaptchaControl/ReCaptcha.php [hunk 1]
```diff
@@ -63,6 +67,9 @@ public function validate($remoteIP, array $post)
 		if (!isset($post[self::RESPONSE_KEY])) {
 			return FALSE;
 		}
+		if (!isset($this->validateCache[$post[self::RESPONSE_KEY]])) {
```
Comment (uestla) [inline src/ReCaptchaControl/ReCaptcha.php]: Shouldn't be here positive condition?

**Comment by uestla** [comment]
I see. Thanks a lot for your patience! :-)

**Comment by uestla** [comment]
Merged, combed a little and released as 4.0.2 :-)
Thanks again!

File: src/ReCaptchaControl/ReCaptcha.php (hunk 1 above)
Comment (Trejjam) [inline src/ReCaptchaControl/ReCaptcha.php]: Yes, it was typo, thanks!

**Comment by uestla** [comment]
@Trejjam I'm still thinking about this - wouldn't it be enough in your use case to just remove the reCAPTCHA from the form before the second validation?

**Comment by Trejjam** [comment]
Probably yes, I do not know why my form is validated twice (I do not call explicitly `validate()`, I will xDebug it later). In the other side, the form could be valid all time (it is still one request).

**Comment by uestla** [comment]
@Trejjam 
The other flaw this update could technically cause (although it is very unlikely) when two different forms are validated in one request, each one of them having reCAPTCHA and in one of them the validation key will be valid and in another one not - but with caching them the same key would be valid in both of them, since one ReCaptcha instance is common for all forms...

I thought you are calling the validation manually and knowingly, if it's happening "behind you back", I would consider resolving this in your application and then maybe revert this update.

Please let me know ;-)

**Comment by Trejjam** [comment]
I think that each control have only one form.

Dne 30. 8. 2016 18:50 napsal uživatel "uestla" <notifications@github.com>:

**Comment by uestla** [comment]
@Trejjam Yes, but the `ReCaptcha` that makes the validation itself is instantiated only once ([see here](https://github.com/uestla/ReCaptchaControl/blob/a6a46b8c3c44c3ee5256e4d0425959c329180be3/src/ReCaptchaControl/ReCaptchaControl.php#L104) - the first parameter).

**Comment by Trejjam** [comment]
Ah, now i see. There is possibility validate (and validate by cache) on two different column.
But:
- send two form with identical value in `g-recaptcha-response` (non sense in PHP & Nette - only one `_do` value is possible)
- it have bigger problem then only one `_do`, because gRecaptcha fill field named `g-recaptcha-response`, so it can be filled only once `g-recaptcha-response[]` != `g-recaptcha-response`. So if you use two reCaptcha input in one form PHP receive only last
- and also this library search for `g-recaptcha-response` that will be all time same (per request)

I miss this [BC break](https://github.com/uestla/ReCaptchaControl/commit/fd920b588e60d8d1f1a6d790a99fa5f3b449dd73) and I `addRule([ReCaptchaControl::class, 'validateValid'])` to control. Firstly I think it is caused by Nette 2.4 (validation problem was reported in that time)

I still think `$responseCache` remove unnecessary problem with repeate form validation

**Comment by uestla** [comment]
@Trejjam I see! So the double-validation problem was caused by the BC break...

I think I could trigger a E_DEPRECATED error when you manually `addRule` the validation so that it won't be so hidden then.

After that I think it will be OK to remove the response caching because every double validation that is done manually and knowingly is something everyone should be thinking through and unset the reCAPTCHA when needed (although I consider this really marginal use case).

Do you agree?

**Comment by Trejjam** [comment]
I am not sure, but it is right.

E_DEPRECATED sound's good

**Comment by uestla** [comment]
Added & tagged as 4.0.3 :-)

Thanks for your time!

**Pull request merged**